src/ocrd/processor/base.py

```diff
@@ -71,31 +73,36 @@ class Processor():
     """
 
     @property
-    def metadata(self):
+    def metadata(self) -> dict:
         """the ocrd-tool.json dict of the package"""
         if hasattr(self, '_metadata'):
             return self._metadata
         self._metadata = json.loads(resource_string(self.__module__.split('.')[0], 'ocrd-tool.json'))
+        report = OcrdToolValidator.validate(self._metadata)
```

Comment:
Does that already expand defaults in `self._metadata` as side effect?

Reply:
Exactly, in addition to the normal validation, `default` is expanded and `deprecated` yields `JsonSchemaDeprecationWarning`s.

Reply:
Fantastic!

Oh, I see, so that's how parameter defaults were _already_ default-extend to date – we always use the JsonValidator in the ParameterValidator.

Did you also try to remove the hard-coded defaults for `resource_locations` and `resource.type` and `resource.parameter_usage` and `resource.path_in_archive` and `resource.version_range` to see if everything still works?

Reply:
> Oh, I see, so that's how parameter defaults were already default-extended to date – we always use the JsonValidator in the ParameterValidator.

Yes, but for the `OcrdToolValidator` used only the non-default-extending JSON Schema validator until now, so we did not expand them.



> Did you also try to remove the hard-coded defaults for `resource_locations` and `resource.type` and `resource.parameter_usage` and `resource.path_in_archive` and `resource.version_range` to see if everything still works?

Not yet, but is upcoming.

Reply:
This will require some significant refactoring. `get_ocrd_tool` cannot be in the `ocrd_utils` package anymore because the default expansion requires `OcrdValidator` from the `ocrd_validators` package and this would create a circular input (due to all the `import` in `__init__.py`) - avoiding that was the original reason for the separate packages.

Reply:
This is even more complicated because it now requires moving dependent methods out of ocrd_utils as well. Should I really go down this rabbit hole?

Reply:
Ouch! No, you're right: this is not worth it. Let's track this in an issue for later.

Reply:
https://github.com/OCR-D/core/tree/validate-ocrd-tool-runtime-refactoring WIP of the refactoring in this branch.